src/util/misc.ts

```diff
@@ -197,56 +219,156 @@ import { cos } from './cos';
+     * @returns {Object[]} array of size n (for miter stroke) or 2n (for bevel or round) of all suspected points. Each element is an object with projectedPoint, originPoint and bisector.
      */
     projectStrokeOnPoints: function (points, options, openPath) {
-      var coords = [], s = options.strokeWidth / 2,
+
+      if (points.length <= 1) { return []; }
+
+      var coords = [],
+          s = options.strokeWidth / 2,
+          scale = new Point(options.scaleX, options.scaleY),
           strokeUniformScalar = options.strokeUniform ?
-            new Point(1 / options.scaleX, 1 / options.scaleY) : new Point(1, 1),
-          getStrokeHatVector = function (v) {
-            var scalar = s / (Math.hypot(v.x, v.y));
-            return new Point(v.x * scalar * strokeUniformScalar.x, v.y * scalar * strokeUniformScalar.y);
-          };
-      if (points.length <= 1) {return coords;}
+            new Point(1 / options.scaleX, 1 / options.scaleY) :
+            new Point(1, 1);
+
+      function scaleHatVector(hatVector, scalar) {
+        return hatVector.multiply(strokeUniformScalar).scalarMultiply(scalar);
+      }
+
       points.forEach(function (p, index) {
         var A = new Point(p.x, p.y), B, C;
         if (index === 0) {
           C = points[index + 1];
-          B = openPath ? getStrokeHatVector(fabric.util.createVector(C, A)).add(A) : points[points.length - 1];
+          B = openPath ? A : points[points.length - 1];
         }
         else if (index === points.length - 1) {
           B = points[index - 1];
-          C = openPath ? getStrokeHatVector(fabric.util.createVector(B, A)).add(A) : points[0];
+          C = openPath ? A : points[0];
         }
         else {
           B = points[index - 1];
           C = points[index + 1];
         }
-        var bisector = fabric.util.getBisector(A, B, C),
-            bisectorVector = bisector.vector,
+        //  safeguard in case `points` are not `Point`
+        B = new Point(B.x, B.y);
+        C = new Point(C.x, C.y);
+
+        if (openPath && (index === 0 || index === points.length - 1)) {
+          var D = index === 0 ? C : B,
+              scaledA = A.multiply(scale),
+              scaledD = D.multiply(scale);
+
+          var vector = fabric.util.createVector(
+                options.strokeUniform ? scaledA : A,
+                options.strokeUniform ? scaledD : D
+              ),
+              hatOrthogonalVector = fabric.util.getOrthogonalUnitVector(vector),
+              orthogonalVector = scaleHatVector(hatOrthogonalVector, s);
+
+          var proj1 = A.add(orthogonalVector),
+              proj2 = A.subtract(orthogonalVector);
+
+          [proj1, proj2].forEach(proj => {
+            coords.push({ 
+              'projectedPoint': proj,
+              'originPoint': A,
+              'bisector': bisector
+            });
+          });
+
+          return;
+        }
+
+        var bisector,
+            scaledA,
```

Comment:
I prefer that from this point we use a single set of variables intead of `A`, `B`, `C` and `scaledA`, `scaledB`, `scaledC`.